### PR TITLE
add a new def-fn linter

### PR DESCRIPTION
**Pull request by andreyorst**
Please answer the following questions and leave the below in as part of your PR.

- [x] I have read the [developer documentation](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md).

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).

- [x] This PR contains a [test](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#tests) to prevent against future regressions

- [x] I have updated the [CHANGELOG.md](https://github.com/clj-kondo/clj-kondo/blob/master/CHANGELOG.md) file with a description of the addressed issue.

Hi and sorry for the delay!

As discussed in https://github.com/clj-kondo/clj-kondo/issues/1920 I've added a new linter for this case. I've added tests for the common cases, and some cases that you've listed in the comments/

### CODE REVIEW

**Comment by borkdude** [comment]
@andreyorst I think `:def-fn` should be disabled by default. This will likely also fix the failing tests.

**Comment by andreyorst** [comment]
> @andreyorst I think `:def-fn` should be disabled by default (only very serious lint issues are enabled by default, since a while, since it would break people's builds if new linters like this were enabled by default). This will likely also fix the failing tests.

Rather than disabling, maybe it should be an `info` level lint?

**Comment by borkdude** [comment]
The policy is that stylistic linters are off by default.

**Comment by borkdude** [comment]
Can you fix the remaining failing tests? This can be done by supplying the explicit config. See other tests for examples.

**Comment by andreyorst** [comment]
> Can you fix the remaining failing tests? This can be done by supplying the explicit config. See other tests for examples.

sure

**Comment by andreyorst** [comment]
Sorry, got a bit messed up by my `~/.config/clj-kondo/config.edn` which included custom hooks, that I've mentioned in the parent issue, which messed with running tests in the project. Wasn't expecting that the user config will be picked up.

**Comment by andreyorst** [comment]
@borkdude, I'm not sure what's up with CircleCI, but everything else seems to work

**Comment by borkdude** [comment]
Thanks! I don't know what's up with CircleCI either, but this happens for some users, whereas for others it doesn't. Very weird. Anyway, I'll have a look later today. Thanks a lot!

**Comment by andreyorst** [comment]
> but this happens for some users, whereas for others it doesn't

May have something to do with this: https://support.circleci.com/hc/en-us/articles/360043679453-CircleCI-Terms-of-Service-Violation-Sanctioned-Country

**Comment by borkdude** [comment]
Looks good, but there's two things left to do:

1. For linters that aren't on by default, we check if they are disabled and then skip the check (to save some cycles). There is a function `utils/linter-disabled?` that you can use for this.

2. Update the `linters.md` page with the documentation for this new linter.

**Comment by andreyorst** [comment]
Done.

What do you think about changing messages from:

- `consider using 'defn' instead of 'fn' inside 'def'`
- `consider using 'defn' inside 'let' instead of 'fn' inside 'let' inside 'def'`

To:

- `consider using 'defn'`
- `consider using 'defn' inside 'let'`

---

Also, in the hooks I've provided in the issue ticket, I've constructed a fixed version of the code, but as far as I know, Clj-kondo doesn't have an API for suggestions. LSP servers often have such a thing that they can fix your code based on some suggestion from the linter, so I was wondering if there's something like that?

**Comment by borkdude** [comment]
> LSP servers often have such a thing that they can fix your code based on some suggestion from the linter

This would have to be done as part of clojure-lsp, let's address that separately after we get this merged

File: src/clj_kondo/impl/analyzer.clj [hunk 1]
```diff
@@ -886,6 +911,8 @@
         varargs-min-arity (when arities (get-in arities [:varargs :min-arity]))
         arglist-strs (when arities (into [] (keep :arglist-str) (vals arities)))
         parsed-bodies (mapcat :parsed parsed-bodies)]
+    (when-let [kind (def-fn? ctx)]
```
Comment (borkdude) [inline src/clj_kondo/impl/analyzer.clj:914]: I would put the linter-disabled? call here so you can avoid the call to `def-fn?` when it's not necessary

**Comment by borkdude** [comment]
About the change in messages. Proposal, just:

```
Use defn instead of def + fn
```

in all cases, without single quotes (these aren't used anywhere else in the messages either).

**Comment by andreyorst** [comment]
Changed messages to:

- `Use defn instead of def + fn`
- `Use let + defn instead of def + let + fn`

Though I don't think `let + defn` conveys the idea clearly enough

**Comment by borkdude** [comment]
I'd say let's keep it simple and have only one message. We can add to the docs how to rewrite `let + def + fn` to `defn`. We can refer to those docs in case people have questions. They will probably read those docs anyway since the linter isn't enabled by default.

**Comment by andreyorst** [comment]
should be good now

**Comment by borkdude** [comment]
Merged via https://github.com/clj-kondo/clj-kondo/commit/51ccf596e42d7a4bc43af87ed2371d74d2059371 - thanks a lot!

**Pull request closed without merging**